allow retrieving proxy image using blob

silvestry opened the pull request:
This PR adds option in runtime to proxy image by retrieving blob. This is driven by security consideration where the proxy endpoint needs to be called via post call, passing csrf token in header.

silvestry committed:
Add option to get proxy image with objectURL

silvestry committed:
Maintain objectURL in DocSafeImage

hedgerwang commented on src/DocsSafeImage.js:
```diff
@@ -172,6 +184,16 @@ class DocsSafeImage extends React.PureComponent {
+  _getProxyImageObjectURL = async (runtime, src): Promise<void> => {
+    try {
+      const imageBlob = await runtime.getProxyImageBlob(src);
+      const objectURL = URL.createObjectURL(imageBlob);
```
This should skip if the component is no longer mounted.

hedgerwang commented on src/DocsSafeImage.js:
```diff
@@ -172,6 +184,16 @@ class DocsSafeImage extends React.PureComponent {
+      const objectURL = URL.createObjectURL(imageBlob);
+      this.setState({objectURL});  
+    } catch(ex) {
+      this.setState({error: true, pending: false});
```
This should skip if the component is no longer mounted.

hedgerwang commented on src/DocsSafeImage.js:
```diff
@@ -47,27 +48,38 @@ class DocsSafeImage extends React.PureComponent {
-    if (src && runtime && runtime.canProxyImageSrc(src)) {
+    let empty = false;
+
+    if (objectURL) {
```
`objectURL` is never updated once it's set. This won't work when `src` changes. 
The component should handle he case when `src` updates.

silvestry replied:
Good catch. I totally missed this one. Will update `componentWillReceiveProps`

silvestry replied:
And also updated `_getProxyImageObjectURL` on async return will check if src is still valid

hedgerwang commented on src/DocsSafeImage.js:
```diff
@@ -47,27 +48,38 @@ class DocsSafeImage extends React.PureComponent {
+    } else if (src && runtime && runtime.canProxyImageSrc(src)) {
       src = runtime.getProxyImageSrc(src);
+    } else if (!src) {
+      empty = true;
```
How is the variable `empty` be used? It does not seem doing anything.

silvestry replied:
This was part of original code on line 118 `return (error || empty) ?` and line 87 where it sets css `docs-safe-image-empty`

hedgerwang commented on src/DocsSafeImage.js:
```diff
@@ -172,6 +184,16 @@ class DocsSafeImage extends React.PureComponent {
+
+  _getProxyImageObjectURL = async (runtime, src): Promise<void> => {
+    try {
+      const imageBlob = await runtime.getProxyImageBlob(src);
```
How would the UI look like while loading the blob?

silvestry replied:
There is a white box. Not ideal, but reasonable?

<img width="528" alt="Screen Shot 2021-05-03 at 9 14 31 PM" src="https://user-images.githubusercontent.com/40640760/116960254-b80ca980-ac54-11eb-814a-f265e060eedc.png">

hedgerwang replied:
lgtm.

silvestry committed:
Check unmount and allow src change

silvestry commented:
@hedgerwang : Thanks for the review. I've updated per your excellent callouts - 1. check for unmount and 2. addressed change in src. Please kindly re-review. Thanks!

hedgerwang commented on src/DocsSafeImage.js:
```diff
@@ -42,32 +43,49 @@ class DocsSafeImage extends React.PureComponent {
+      if (objectURL) {
+        src = objectURL;
+      } else if (src && runtime && runtime.canProxyImageBlob(src)) {
+        this._getProxyImageObjectURL(runtime, src);
```
Ideally, `render` should not have side-effect.
Instead, using life-cycle method to manage side-effects.

```
componentDidMount(): {
  this._maybeLoadObjectUrl();
}

componentDidUpdate(): {
  this._maybeLoadObjectUrl();
}

_maybeLoadObjectUrl() {
  const {src} = this.props;
  const {objectURL} = this.state;
  const {runtime} = this.context.docsContext;
  if (!objectURL && src && runtime && runtime.canProxyImageBlob(src)) {
    this._getProxyImageObjectURL(runtime, src);
  }  
}
```

silvestry committed:
Move side-effect to componentDid*
